### PR TITLE
feature: support discovery center

### LINKED ISSUE #1279: feature: eureka integration
This is an idea about integrating APISIX with eureka：

1. Added  discovery component, such as eureka;
2. Use the `upstream.service_name`  to associate the application name (`spring.application.name` in spring boot)
3. In `apisix.balancer`, according to the `upstream.service_name`, first find the application nodes in the discovery;

If you want to register APISIX with the registry, it will be more complicated. Do you have any requirements for this?

### PR DESCRIPTION
**Pull request by qiujiayu**
Fix #1279

### CODE REVIEW

File: apisix/http/service.lua [hunk 1]
```diff
@@ -24,6 +26,48 @@ local _M = {
 }
 
 
+local function filter(service)
```
Comment (membphis) [inline apisix/http/service.lua:29]: https://github.com/apache/incubator-apisix/pull/1402

we can merge this PR first

Comment (membphis) [reply apisix/http/service.lua]: @qiujiayu you can rebase your branch with master, #1402 has been merged.

Comment (qiujiayu) [reply apisix/http/service.lua]: done

**Comment by qiujiayu** [comment]
@membphis @moonming review please.

File: apisix/core/table.lua [hunk 2]
```diff
@@ -25,13 +25,14 @@ local type         = type
...
 local _M = {
-    version = 0.1,
+    version = 0.2,
     new     = new_tab,
     clear   = require("table.clear"),
     nkeys   = nkeys,
     insert  = table.insert,
     concat  = table.concat,
     clone   = require("table.clone"),
+    isarray = require("table.isarray"),
```
Comment (moonming) [inline apisix/core/table.lua:35]: why add this line?

File: apisix/discovery/eureka.lua [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+local ngx                = ngx
+local ngx_timer_at       = ngx.timer.at
+local ngx_timer_every    = ngx.timer.every
+local string_sub         = string.sub
+local string_find        = string.find
+local log                = core.log
+
+local default_weight
+local applications
+local useragent = 'ngx_lua-apisix/v' .. core.version.VERSION
```
Comment (moonming) [inline apisix/discovery/eureka.lua:36]: why add this UA?

File: apisix/discovery/eureka.lua [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+                local nodes = up_apps[app.name]
+                if not nodes then
+                    nodes = core.table.new(#app.instance, 0)
+                    up_apps[app.name] = nodes
+                end
+                core.table.insert(nodes, {
+                    host = ip,
+                    port = port,
+                    weight = metadata and metadata.weight or default_weight,
+                    metadata = metadata,
```
Comment (moonming) [inline apisix/discovery/eureka.lua:207]: we need to remove `metadata` as discussed in mailing list.

File: apisix/discovery/eureka.lua [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    end
+
+    local ok, err = core.schema.check(schema, local_conf.eureka)
+    if not ok then
+        error("invalid eureka configuration: " .. err)
+        return
+    end
+    default_weight = local_conf.eureka.weight or 100
+    ngx_timer_at(0, fetch_full_registry)
+    ngx_timer_every(30, fetch_full_registry)
```
Comment (moonming) [inline apisix/discovery/eureka.lua:243]: why two timer?

File: apisix/discovery/eureka.lua [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+local error              = error
+local ngx                = ngx
+local ngx_timer_at       = ngx.timer.at
+local ngx_timer_every    = ngx.timer.every
+local string_sub         = string.sub
+local string_find        = string.find
+local log                = core.log
+
+local default_weight
+local applications
```
Comment (moonming) [inline apisix/discovery/eureka.lua:35]: why top level args?

File: apisix/router.lua [hunk 7]
```diff
@@ -88,4 +108,8 @@ function _M.http_routes()
 end
 
 
+-- for test
+_M.filter = filter
```
Comment (moonming) [inline apisix/router.lua:112]: for test?

File: t/admin/upstream.t [hunk 8]
```diff
@@ -235,7 +235,7 @@ GET /t
 GET /t
 --- error_code: 400
 --- response_body
-{"error_msg":"invalid configuration: property \"nodes\" validation failed: expect object to have at least 1 properties"}
+{"error_msg":"invalid configuration: property \"nodes\" validation failed: object matches none of the requireds"}
```
Comment (moonming) [inline t/admin/upstream.t:238]: I don't think it should be modified here.

File: t/admin/upstream.t [hunk 9]
```diff
@@ -523,7 +523,7 @@ GET /t
 GET /t
 --- error_code: 400
 --- response_body
-{"error_msg":"invalid configuration: property \"nodes\" validation failed: failed to validate 127.0.0.1:8080 (matching \".*\"): wrong type: expected integer, got string"}
+{"error_msg":"invalid configuration: property \"nodes\" validation failed: object matches none of the requireds"}
```
Comment (moonming) [inline t/admin/upstream.t:526]: ditto

File: t/admin/upstream.t [hunk 10]
```diff
@@ -553,7 +553,7 @@ GET /t
 GET /t
 --- error_code: 400
 --- response_body
-{"error_msg":"invalid configuration: property \"nodes\" validation failed: failed to validate 127.0.0.1:8080 (matching \".*\"): expected -100 to be greater than 0"}
+{"error_msg":"invalid configuration: property \"nodes\" validation failed: object matches none of the requireds"}
```
Comment (moonming) [inline t/admin/upstream.t:556]: ditto

File: apisix/router.lua (hunk 7 above)
Comment (qiujiayu) [reply apisix/router.lua:112]: be used in `t/admin/balancer.t`

File: apisix/discovery/eureka.lua (hunk 4 above)
Comment (qiujiayu) [reply apisix/discovery/eureka.lua:207]: https://github.com/apache/incubator-apisix/pull/1440/files/2a2da03a7b71ad3314a89253fa478182fa66c22f#diff-98c27e6d94725a36de10557d23b1ca94R211

remove in here

File: apisix/core/table.lua (hunk 2 above)
Comment (qiujiayu) [reply apisix/core/table.lua:35]: https://github.com/apache/incubator-apisix/pull/1440/files/2a2da03a7b71ad3314a89253fa478182fa66c22f#diff-e3e08c15cae0b9924d16c46ac8f64230R345

https://github.com/apache/incubator-apisix/pull/1440/files/2a2da03a7b71ad3314a89253fa478182fa66c22f#diff-c96e5aa165284ada8a16fee994fec508R54

https://github.com/apache/incubator-apisix/pull/1440/files/2a2da03a7b71ad3314a89253fa478182fa66c22f#diff-5e074f24e0cf98f1f68e6f4390a49164R38

Check that table is a array.

File: apisix/discovery/eureka.lua (hunk 3 above)
Comment (qiujiayu) [reply apisix/discovery/eureka.lua:36]: The effect is very small, can be deleted

File: t/admin/upstream.t (hunk 8 above)
Comment (qiujiayu) [reply t/admin/upstream.t:238]: If not changed, the case cannot be passed.

or we need change `jsonschema`'s message

File: apisix/discovery/eureka.lua (hunk 5 above)
Comment (qiujiayu) [reply apisix/discovery/eureka.lua:243]: Once APISIX is started,  needs fetch data immediately. so use `ngx_timer_at(0, fetch_full_registry)`.

Then every 30 seconds for a time.

File: .travis/linux_openresty_runner.sh [hunk 11]
```diff
@@ -102,6 +102,13 @@ do_install() {
         tar -xvf grpcurl-amd64.tar.gz
         mv grpcurl build-cache/
     fi
+
+    if [ ! -f "build-cache/eureka" ]; then
+        wget https://github.com/api7/eureka-for-test/releases/download/v1.9.8/eureka.tar.gz
```
[the remark this answers is not in the document]
Comment (qiujiayu) [reply .travis/linux_openresty_runner.sh]: find a third party image.

File: apisix/discovery/eureka.lua (hunk 6 above)
Comment (qiujiayu) [reply apisix/discovery/eureka.lua:35]: `applications` use to save the data from eureka.

File: apisix/discovery/eureka.lua (hunk 3 above)
Comment (qiujiayu) [reply apisix/discovery/eureka.lua]: deleted

**Comment by qiujiayu** [comment]
> There are too few test cases.

Some test cases have been added.

File: apisix/balancer.lua [hunk 12]
```diff
@@ -42,30 +44,37 @@ local lrucache_checker = core.lrucache.new({
...
 
 
 local function fetch_health_nodes(upstream, checker)
+    local nodes = upstream.nodes
     if not checker then
-        return upstream.nodes
+        local new_nodes = core.table.new(0, #nodes)
+        for _, node in ipairs(nodes) do
+            -- TODO filter with metadata
+            new_nodes[core.table.concat({node.host, ":", node.port})] = node.weight
```
[the remark this answers is not in the document]
Comment (qiujiayu) [reply apisix/balancer.lua]: Done

File: apisix/discovery/eureka.lua [hunk 13]
```diff
@@ -157,7 +157,7 @@ local function parse_instance(instance)
     local ip = instance.ipAddr
     if not ipmatcher.parse_ipv4(ip) and
             not ipmatcher.parse_ipv6(ip) then
-        log.error("invalid ip:", ip)
+        log.error(instance.app, " service ", instance.hostName, " node IP ", ip, " is invalid(must be IPv4 or IPv6).")
```
Comment (membphis) [inline apisix/discovery/eureka.lua]: this line is too long. 
you can rebase your branch with master branch and run `make lint`, it show you similar warning message.

**Pull request merged**